torch/distributed/_sharded_tensor/api.py

```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._barrier(rpc_workers)
 
+    @classmethod
+    def init_from_local_shards(
+        cls,
+        local_shards: List[torch.Tensor],
+        shards_metadata: List[ShardMetadata],
+        process_group=None
+    ):
+
+        if len(local_shards) == 0 or len(shards_metadata) == 0:
+            raise ValueError("local_shards or shards_metadata must not be empty!")
+
+        sharded_tensor = cls.__new__(cls)
+
+        # prepare initialization
+        sharded_tensor._prepare_init(process_group=process_group)
+
+        sharded_tensor._sharding_spec = EnumerableShardingSpec(shards_metadata)
+        sharded_tensor._sharding_metadata = shards_metadata
+
+        current_rank = dist.get_rank(sharded_tensor._process_group)
+
+        local_shard_idx = 0
+        num_dims = len(shards_metadata[0].shard_lengths)
+        offset_length_tuple_list_by_dim: List[List[Tuple[int, int]]] = [[] for _ in range(num_dims)]
+        # validate the metadata and add to local_shards
+        for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_metadata.placement)
+            # record shard_offsets/lengths for dims recovery later
+            shard_offsets = shard_metadata.shard_offsets
+            shard_lengths = shard_metadata.shard_lengths
+
+            for i, (offset, length) in enumerate(zip(shard_offsets, shard_lengths)):
+                offset_length_tuple_list_by_dim[i].append((offset, length))
+
+            # validate local_shards and metadata
+            if current_rank == rank:
+                local_shard = local_shards[local_shard_idx]
+                if shard_metadata.shard_lengths != list(local_shard.size()):
+                    raise ValueError(
+                        f'Local shard tensor is incompatible with local ShardMetadata! '
+                        f'local shard tensor size: {local_shard.size()}, '
+                        f'local ShardMetadata shard lengths: {shard_metadata.shard_lengths}'
+                    )
+
+                sharded_tensor._local_shards.append(Shard(local_shard, shard_metadata))
+                local_shard_idx += 1
+
+        if local_shard_idx != len(local_shards):
+            raise RuntimeError(f'Number of local shard tensors does not match '
+                               f'number of ShardMetadata on rank {current_rank}')
+
+        # recover the global dims information
+        dims = [0] * num_dims
+        for i, offset_length_tuple_list in enumerate(offset_length_tuple_list_by_dim):
```

Comment:
Don't we need to do somme validation that there are no "gaps" between the shards? I'm wondering if we should add `ShardedTensorMetadata` as part of this change itself since it would be much more clear in terms of validation.

For example, ShardedTensorMetadata would contain the overall_dims of the tensor and that can be easily verified with shards_metadata.

Reply:
yeah I think this make sense, if we have ShardedTensorMetadata, we only need to verify the dims constructed by List[ShardedMetadata] matches overall_dims. In terms of "gaps" validation, we do have it checked here and raise error when `dims[i] < offset`